### PR TITLE
(WIP) [inductor][foreach] Horizontally fuse concat inputs that require copy kernels

### PR DESCRIPTION
**Pull request by ColinPeppler**
Most Inductor IR will require a copy kernel (i.e. load + store) if it's an input of a `ConcatKernel`. The only exeception is when the input is a `Pointwise` or `Reduction`. In this case, just ask the input to write directly to the concat's output buffer.

Before this PR: Create a separate copy kernel for each input.

With this PR: Horizontally fuse copy kernels when there's more than 10 inputs.
* Note: fusing with fewer inputs will reduce launch overhead but might prohibit better fusion opportunities (?).

```
$ python test_foreach.py -k test/inductor/fuse_concat 
----------------------------------------------------------------------
Ran 3 tests in 17.468s

OK
```

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #115217

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

Differential Revision: [D52423531](https://our.internmc.facebook.com/intern/diff/D52423531)

### CODE REVIEW

File: torch/_inductor/ir.py [hunk 1]
```diff
@@ -3426,21 +3426,21 @@ def create(cls, inputs, dim):
...
+            if isinstance(input.data, BaseView):
+                input_unwrapped = input.data.unwrap_view()
             else:
-                input_unwrapped = inputs[i].data
+                input_unwrapped = input.data
 
             if (
-                input_unwrapped.is_input_buffer()
-                and inputs[i].get_device().type == "cuda"
+                copy_required and len(inputs) >= 10
```
Comment (chenyang78) [inline torch/_inductor/ir.py:3442]: nit - perhaps add some comment to explain why we are doing this? Maybe something similar to the PR summary?

File: test/inductor/test_foreach.py [hunk 2]
```diff
@@ -548,6 +548,42 @@ def fn(x1, x2, x3, w1, w2, w3):
...
+        # Case: num-operands ≥ 10, create a foreach node.
+        self.check_model_cuda(fn, (x, w, 10))
+        self.assertEqual(torch._inductor.metrics.generated_kernel_count, 3)
+
+    @requires_cuda()
+    @torch._inductor.config.patch("max_pointwise_cat_inputs", 0)
+    def test_fuse_concat_non_input_views(self):
+        def fn(x, slice_sz):
+            x = x + 1  # Make it pointwise to avoid views on inputs.
+            views = torch.ops.aten.split.Tensor(x, slice_sz, dim=0)
```
Comment (aakhundov) [inline test/inductor/test_foreach.py:574]: Nit: can just use `torch.split` here instead? Like here:

https://github.com/pytorch/pytorch/blob/f6dfbffb3bb46ada6fe66b5da4f989f9d4d69b3c/test/dynamo/test_functions.py#L2462

File: test/inductor/test_foreach.py [hunk 3]
```diff
@@ -548,6 +548,42 @@ def fn(x1, x2, x3, w1, w2, w3):
...
+            return torch.ops.aten.cat(views)
```
Comment (aakhundov) [inline test/inductor/test_foreach.py:575]: Is there a reason we're not using `torch.cat` here? If so, please add a comment.

File: torch/_inductor/ir.py [hunk 4]
```diff
@@ -3464,7 +3464,7 @@ def can_realize_into_without_copy(cls, src):
         )
 
     @classmethod
-    def realize_into(cls, src, dst):
+    def realize_into(cls, src, dst, with_copy=False):
```
Comment (aakhundov) [inline torch/_inductor/ir.py:3467]: I don't quite understand the purpose of the new `with_copy=True` parameter. Seems that it's not used in the function's body except being propagated, forced to `True` at some places, and returned as an additional output value. And it's not set from the modified external code. Can we, therefore, just add the additional output and return either `True` or `False` depending on what is happening inside the function? Or am I missing something?

**Comment by aakhundov** [comment]
@ColinPeppler Please lease fix the linter, too.

**Comment by ColinPeppler** [comment]
@ColinPeppler has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D52423531).

File: torch/_inductor/ir.py (hunk 4 above)
Comment (ColinPeppler) [reply torch/_inductor/ir.py:3467]: Yes, actually your idea to explicitly set the value for `with_copy` when we return is clearer.

**Pull request closed without merging**